src/k8s/pkg/k8s/setup/serviceargs.go

```diff
@@ -2,19 +2,44 @@ package setup
 
 import (
 	"fmt"
+	"os"
+	"regexp"
+	"strings"
 
+	apiv1 "github.com/canonical/k8s/api/v1"
+	"github.com/canonical/k8s/pkg/snap"
 	"github.com/canonical/k8s/pkg/utils"
 )
 
+const argumentPattern = `--?($ARGUMENT_NAME)[[:space:]]*=[[:space:]]*([^[:space:]\n]+)`
+
 // InitServiceArgs handles the setup of services arguments.
-//   - For each service, copies the default arguments files from the snap under $SNAP_DATA/args
+//   - For each service, copies the default arguments files from the snap under $SNAP_DATA/args and apply any overwrites
 //   - Note that the `k8sd` service is already configured in the snap install hook and thus not included here
-func InitServiceArgs() error {
+func InitServiceArgs(overwrites *apiv1.ExtraServiceArgs) error {
+	if overwrites == nil {
+		overwrites = &apiv1.ExtraServiceArgs{}
+	}
+
 	for _, service := range []string{"containerd", "k8s-dqlite", "kube-apiserver", "kube-controller-manager", "kube-proxy", "kube-scheduler", "kubelet"} {
-		err := utils.CopyFile(utils.SnapPath("k8s/args", service), utils.SnapDataPath("args", service))
+		serviceArgs, err := utils.ReadFile(snap.Instance().Path("k8s/args", service))
 		if err != nil {
-			return fmt.Errorf("failed to copy %s args: %w", service, err)
+			return fmt.Errorf("failed to read %s args: %w", service, err)
 		}
+
+		// Apply overwrites for each service
+		if ao, exists := (*overwrites)[service]; exists {
+			for argument, value := range ao {
+				pattern := strings.ReplaceAll(argumentPattern, "$ARGUMENT_NAME", argument)
+				re := regexp.MustCompile(pattern)
+				if matches := re.FindStringSubmatch(serviceArgs); matches != nil {
+					argumentWithNewValue := strings.ReplaceAll(matches[0], matches[2], value)
+					serviceArgs = strings.ReplaceAll(serviceArgs, matches[0], argumentWithNewValue)
+				}
+			}
+		}
```

Comment:
I see the value in having a regex for this, but if we do so, please ensure we have proper unit tests for the cases that we _at least_ expect to happen most of the time.

As reference or example, see https://github.com/canonical/microk8s-cluster-agent/blob/main/pkg/snap/services.go and https://github.com/canonical/microk8s-cluster-agent/blob/main/pkg/snap/services_test.go

Also, as a general comment, I would prefer working on structured data (e.g. arguments should be a `map[string]string` or something similar when we mutate them), and only ever render them at the end (in the format of the args file we have).

Otherwise, we mix business logic with the representation, and makes it hard to test/validate/maintain in the future.

Reply:
Very good point. I parse the argument file now into a map and overwrite the values there. Way cleaner